File: 2022/go/day02.go
package main

import (
	"aoc/utils"
	"fmt"
	"strings"
)

func getWinner(play string) string {
	switch play {
	case "rock":
		return "paper"
	case "paper":
		return "scissors"
	case "scissors":
		return "rock"
	}
	return "--"
}

func getLoser(play string) string {
	switch play {
	case "rock":
		return "scissors"
	case "paper":
		return "rock"
	case "scissors":
		return "paper"
	}
	return "--"
}

func tranlate_opp_play(play string) string {
	switch play {
	case "A":
		return "rock"
	case "B":
		return "paper"
	case "C":
		return "scissors"
	}
	return "--"
}

func part1_you_play(play string) string {
	switch play {
	case "X":
		return "rock"
	case "Y":
		return "paper"
	case "Z":
		return "scissors"
	}
	return "--"
}

func part2_you_play(opp_play, play string) string {
	if play == "Y" {
		return opp_play
	}

	switch play {
	case "X":
		return getLoser(opp_play)
	case "Y":
		return opp_play
	case "Z":
		return getWinner(opp_play)
	}
	return "--"
}

func part1_score(opp, you string) int {
	opp_play := tranlate_opp_play(opp)
	you_play := part1_you_play(you)

	return pointsForPlay(you_play) + pointsForGame(opp_play, you_play)
}

func part2_score(opp, you string) int {
	opp_play := tranlate_opp_play(opp)
	you_play := part2_you_play(opp_play, you)

	return pointsForPlay(you_play) + pointsForGame(opp_play, you_play)
}

func pointsForGame(opp, you string) int {
	if opp == you {
		return 3
	}
	if opp == "rock" && you == "paper" {
		return 6
	} else if opp == "paper" && you == "scissors" {
		return 6
	} else if opp == "scissors" && you == "rock" {
		return 6
	}
	return 0
}

func pointsForPlay(p string) int {
	switch p {
	case "rock":
		return 1
	case "paper":
		return 2
	case "scissors":
		return 3
	}
	return 0
}

func main() {
	input := utils.ReadPiped()

	part1_points := 0
	part2_points := 0
	for _, line := range input {
		game := strings.Split(line, " ")

		part1_points += part1_score(game[0], game[1])
		part2_points += part2_score(game[0], game[1])
	}

	fmt.Println("Part 1:", part1_points)
	fmt.Println("Part 2:", part2_points)
}

File: 2022/cc/day02.cc
#include "utils/file_reader.h"
#include <string>
#include <vector>
#include <iostream>

enum Play
{
  kRock,
  kPaper,
  kScissors,
  kInvalid,
};

Play getWinner(Play pos)
{
  switch (pos)
  {
  case kRock:
    return kPaper;
  case kPaper:
    return kScissors;
  case kScissors:
    return kRock;
  default:
    return pos;
  }
}

Play getLoser(Play pos)
{
  switch (pos)
  {
  case kRock:
    return kScissors;
  case kPaper:
    return kRock;
  case kScissors:
    return kPaper;
  default:
    return pos;
  }
}

Play translateOpp(char play)
{
  if (play == 'A')
  {
    return kRock;
  }
  if (play == 'B')
  {
    return kPaper;
  }
  if (play == 'C')
  {
    return kScissors;
  }
  return kInvalid;
}

Play translateYou_part1(char play)
{
  if (play == 'X')
  {
    return kRock;
  }
  if (play == 'Y')
  {
    return kPaper;
  }
  if (play == 'Z')
  {
    return kScissors;
  }
  return kInvalid;
}

Play translateYou_part2(Play opp, char you)
{
  if (you == 'Y')
  {
    return opp;
  }
  if (you == 'X')
  {
    return getLoser(opp);
  }
  return getWinner(opp);
}

int pointsForGame(Play opp, Play you)
{
  if (opp == you)
  {
    return 3;
  }
  if (opp == kRock && you == kPaper)
  {
    return 6;
  }
  if (opp == kPaper && you == kScissors)
  {
    return 6;
  }
  if (opp == kScissors && you == kRock)
  {
    return 6;
  }
  return 0;
}

int pointsForPlay(Play p)
{
  switch (p)
  {
  case kRock:
    return 1;
  case kPaper:
    return 2;
  case kScissors:
    return 3;
  default:
    return 0;
  }
}

int part1_score(char opp, char you)
{
  auto opp_play = translateOpp(opp);
  auto you_play = translateYou_part1(you);

  return pointsForPlay(you_play) + pointsForGame(opp_play, you_play);
}

int part2_score(char opp, char you)
{
  auto opp_play = translateOpp(opp);
  auto you_play = translateYou_part2(opp_play, you);

  return pointsForPlay(you_play) + pointsForGame(opp_play, you_play);
}

int main(int argc, char **argv)
{
  auto input = brooks::utils::read(argv[1]);

  int p1 = 0;
  int p2 = 0;
  for (auto line : input)
  {
    p1 += part1_score(line[0], line[2]);
    p2 += part2_score(line[0], line[2]);
  }

  std::cout << "Part 1: " << p1 << std::endl;
  std::cout << "Part 2: " << p2 << std::endl;

  return 0;
}
